src/search.cpp

```diff
@@ -1458,19 +1457,22 @@ Value Search::Worker::qsearch(Position& pos, Stack* ss, Value alpha, Value beta,
     // Note that unlike regular search, which stores literal depth, in QS we only store the
     // current movegen stage. If in check, we search all evasions and thus store
     // DEPTH_QS_CHECKS. (Evasions may be quiet, and _CHECKS includes quiets.)
-    ttDepth = ss->inCheck || depth >= DEPTH_QS_CHECKS ? DEPTH_QS_CHECKS : DEPTH_QS_NORMAL;
+    Depth qsDepth = ss->inCheck || depth >= DEPTH_QS_CHECKS ? DEPTH_QS_CHECKS : DEPTH_QS_NORMAL;
 
     // Step 3. Transposition table lookup
     posKey  = pos.key();
-    tte     = tt.probe(posKey, ss->ttHit);
-    ttValue = ss->ttHit ? value_from_tt(tte->value(), ss->ply, pos.rule50_count()) : VALUE_NONE;
-    ttMove  = ss->ttHit ? tte->move() : Move::none();
-    pvHit   = ss->ttHit && tte->is_pv();
+    auto [ttHit, ttData, ttWriter] = tt.probe(posKey);
+    auto [ttMove, ttValue, ttEval, ttDepth, ttBound, ttIsPv] = ttData; // completely optional to unstructure or not
+    // Need further processing of the saved data
+    ss->ttHit = ttHit;
+    ttMove  = ss->ttHit ? ttMove : Move::none();
+    ttValue = ss->ttHit ? value_from_tt(ttValue, ss->ply, pos.rule50_count()) : VALUE_NONE;
+    pvHit   = ss->ttHit && ttIsPv;
 
     // At non-PV nodes we check for an early TT cutoff
-    if (!PvNode && tte->depth() >= ttDepth
-        && ttValue != VALUE_NONE  // Only in case of TT access race or if !ttHit
```

Comment:
keep comment

Reply:
but there can no longer be any access races after `probe` returns. therefore, it can only be `if (!ttHit)`, in which case the code speaks for itself

Reply:
there can be races while exposing the data, that's still a possibility that's why we add those to the excluded races in the sanitizer. so a one entry while getting copied is still racy

Reply:
the TT is affected by races, period. Any consumer needs to be able to deal with that, that reminder is thus useful.